pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,223 @@
+	localProfileURL := fmt.Sprintf("http://%s", localProfileServer)
+	u, err := url.Parse(localProfileURL)
+	if err != nil {
+		return "", fmt.Errorf("url.Parse: %s", err)
+	}
+	if u.Port() == "" {
+		localProfileURL = fmt.Sprintf("%s:%s", localProfileURL, defaultLocalProfileServerPort)
+	}
+	return fmt.Sprintf("%s/api/v1/local_profile", localProfileURL), nil
+}
+
+// Run a periodic fetch of the currentProfile
+func localProfileTimerTask(handleChannel chan interface{},
+	getconfigCtx *getconfigContext) {
+
+	ctx := getconfigCtx.zedagentCtx
+	iteration := 0
+	localProfileServer := getconfigCtx.localProfileServer
+	if localProfileServer != "" {
+		localProfileURL, err := getLocalProfileURL(localProfileServer)
+		if err != nil {
+			log.Errorf("localProfileTimerTask: getLocalProfileURL: %s", err)
+		} else {
+			if err := getLocalProfileConfig(localProfileURL, iteration, getconfigCtx); err != nil {
+				log.Errorf("localProfileTimerTask: %s", err)
+			}
+		}
+		publishZedAgentStatus(getconfigCtx)
+	}
+
+	// use ConfigInterval as localProfileInterval
+	localProfileInterval := ctx.globalConfig.GlobalValueInt(types.ConfigInterval)
+	interval := time.Duration(localProfileInterval) * time.Second
+	max := float64(interval)
+	min := max * 0.3
+	ticker := flextimer.NewRangeTicker(time.Duration(min),
+		time.Duration(max))
+	// Return handle to caller
+	handleChannel <- ticker
+
+	wdName := agentName + "currentProfile"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	for {
+		select {
+		case <-ticker.C:
+			localProfileServer = getconfigCtx.localProfileServer
+			if localProfileServer != "" {
+				start := time.Now()
+				iteration++
+				localProfileURL, err := getLocalProfileURL(localProfileServer)
+				if err != nil {
+					log.Errorf("getLocalProfileURL: %s", err)
+				} else {
+					if err := getLocalProfileConfig(localProfileURL, iteration, getconfigCtx); err != nil {
+						log.Errorf("localProfileTimerTask getLocalProfileConfig: %s", err)
+					}
+				}
+				ctx.ps.CheckMaxTimeTopic(wdName, "getLocalProfileConfig", start,
+					warningTime, errorTime)
+				publishZedAgentStatus(getconfigCtx)
+			}
+
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func validateAndSetLocalProfile(localProfileBytes []byte, getconfigCtx *getconfigContext) error {
+	var localProfile = &profile.LocalProfile{}
+	err := proto.Unmarshal(localProfileBytes, localProfile)
+	if err != nil {
+		return fmt.Errorf("validateAndSetLocalProfile: Unmarshalling failed: %v", err)
+	}
+	if localProfile.GetServerToken() != getconfigCtx.profileServerToken {
+		// send something to ledmanager ??
+		return fmt.Errorf("validateAndSetLocalProfile: missamtch ServerToken for loacl profile server")
+	}
+	getconfigCtx.currentProfile = localProfile.GetLocalProfile()
```

Comment:
Presumably want some log.Notice if this changes the currentProfile.

Reply:
Modified